### ISSUE TITLE
3.6.0: significant video/audio desync in "Zniw Adventure" during cutscenes

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1499]
**Describe the bug**
As [reported before](https://github.com/adventuregamestudio/ags/issues/1464#issuecomment-1001262805), when running "Zniw Adventure" with the current 3.6.0 engine, and any of the in-game videos are playing, the video may play significantly slower than the sound, or stop animating completely for a while, then jump forward. Meanwhile the audio seem to play fine.
This works well when running 3.5.1 engine (game itself is made with 3.5.0 iirc).

This may or not be a universal issue with playing videos in 3.6.0 (need to be tested).

**Game**
As far as I know it's available only on Steam at the moment:
https://store.steampowered.com/app/904750/Zniw_Adventure/

EDIT: I made this minimal game for easier testing:
[test--351video.zip](https://github.com/adventuregamestudio/ags/files/7868425/test--351video.zip)
But you have to place a video file called `intro_a.ogv` in the game's dir. I am not certain if it's allowed to redistribute the parts of the original game, so don't include it. You may use your own video with audio.

NOTE: the game is made to run at 85 fps, which may or not be relevant.

**To Reproduce**
1. Run the game, select any language
2. In main menu choose "Play" then "Begin new game", then any option for "autosave".
3. An intro video will start playing. Continue watching for around 30 seconds, it's enough to notice the issue.

**Expected behavior**
Video and audio should be playing in sync, video should not stuck.

**UPDATE:**
Commits that may be related:
00627fdd7dfb67f736041429fdc071bf611245b5

**Comment by ivan-mogilko** [comment]
UPDATE: this happens only if the video is playing with its own audio stream; if the audio stream is disabled (by passing necessary flags to PlayVideo), the video seem to play with the proper speed, at least comparing 3.5.1 and 3.6.0 games running in parallel.

There's commit 00627fd that updated apeg library to use SDL2; my assumption is that the new code breaks the sync between audio and video, so I'd focus on researching it now.
What I don't like is that it does not use engine's common audio system, but has everything coded in place (well, it's been always like that, afaik, except previously it used allegro drivers, now it's using OpenAL + SDL).

**Comment by ericoporto** [comment]
> What I don't like is that it does not use engine's common audio system

You mean like an audio channel? If it did, it would have an audio type and the functions to adjust things (like volume) by audio type would work with the video?

**Comment by ivan-mogilko** [comment]
> > What I don't like is that it does not use engine's common audio system
> 
> You mean like an audio channel? If it did, it would have an audio type and the functions to adjust things (like volume) by audio type would work with the video? (like `Game.SetAudioTypeVolume(eVideo, 50)`)

I was refering to the internal process: the audio thread, the slots and so on, what we have in `audio_core.cpp`.
Exposing this to the script would be something else, and if we do I'd rather wish we aim at a more expanded API.

**Comment by ivan-mogilko** [comment]
Hm, I think I found the culprit.

The sync between video and audio in APEG library is done based on audio sample position.
The OpenAl implementation we're using returns current playback position relative to the current buffer, resetting it to zero with each complete buffer or buffer queue (buffer is quite small).
In our audio_core we account for that, and accumulate processed lengths ourselves: 
https://github.com/adventuregamestudio/ags/blob/v.3.6.0.13/Engine/media/audio/openaldecoder.cpp#L105
and use this accumulated length when reporting playback position:
https://github.com/adventuregamestudio/ags/blob/v.3.6.0.13/Engine/media/audio/openaldecoder.cpp#L366

The new implementation in APEG does not have same solution, instead it uses raw values from OpenAL, which causes reported audio position to "jump" back every now and then, which in turn makes video stutter.

(this is a good example of why it's best to use same implementation for everything and not duplicate things...)

**Comment by ivan-mogilko** [comment]
As a small update: i found the fix, theoretically, but also found more strange things while testing other video examples, so will postpone doing any commits.

**Comment by ericoporto** [comment]
Not sure how feasible it would be, but if where the apeg library has to use openal (the `_apeg_audio_*()` functions I think), used callbacks instead, the functions that pass the relevant information could be set on the engine side at init time (to not mix engine code for the audio_core inside of it).

**Comment by ivan-mogilko** [comment]
Pushed a fix: 2fcb329c6ffa65f84641824041744df5a6818236
Details are the commit's description. Besides the problem mentioned above, there was another issue with mojoAL lying about its buffer contents... because it secretly stores everything in float32 (and OGV audio is 16-bit).
Which probably means that our common audio Position property is also calculated incorrectly for non-32bit sounds (going to test that and fix separately if i'm right).
Also I might mention this issue in mojoAl's repository, even though I am not sure if this may be considered a bug, or we're just calculating the audio progress incorrectly. Maybe there are more reliable methods to do this (i'm going to look into this later).

"Zniw"'s videos are working correctly now.
But I'm not closing this yet, because I found more problems with OGVs from other games. They looks similar, but were not fixed by this commit, so I might be missing something else.

**Comment by ivan-mogilko** [comment]
Wasted to much time on this..... in the end I realized that after changing to SDL2/OpenAL APEG detects audio ahead of video most of the time for certain ogvs. This leads to video frames getting "dropped", when it tries to catch up, but in reality it never does.

So I just disabled framedrop: 6c10e6e5e16ec70f5d5986b23217dc5ebf13a70c. Tested with 3 ogvs from 3 different games, and they seem to work and sync well now.
